### PR TITLE
QA: stop nodes after RPC tests, even with --nocleanup

### PR DESCRIPTION
**Pull request by dexX7**
`--nocleanup` should provide a way to preserve test data, but should not have an impact on whether nodes are to be stopped after the test execution.

In particular, when currently running RPC tests with `--nocleanup`, then it may result in several active `bitcoind`, which are not stopped properly.

~~Shutdown errors are ignored, given that a failure likely relates to already terminated processes, caused during the actual testing.~~

Edit:
Further, with `--noshutdown`, the nodes are not stopped explicitly. `--noshutdown` implies `--nocleanup`, to prevent removing datadirs, which are still in use.

### CODE REVIEW

**Comment by laanwj** [comment]
I was also surprised by the bitcoinds still running when specifying `--nocleanup`, it is not very intuitive. Though I'm sure that there can be reasons to want the nodes to remain running, for example to investigate their state (which is more difficult postmortem)

**Comment by dexX7** [comment]
> Though I'm sure that there can be reasons to want the nodes to remain running ... which is more difficult postmortem

There could be a `--noshutdown` option to explicitly prevent shutting down the nodes (which should imply `--nocleanup` to avoid removing datadirs that are still in use), but I think there would probably be not much use for it, due to the seemingly random selection of RPC ports.

Assuming there were such an option, additional reporting of the ports, or alternatively providing options to specify ports, or maybe even an option to run the QT client instead of bitcoind, would be helpful.

I'm not sure how other people use, or want to use, the tests, so it would be interesting to learn more about potential use cases. :)

**Comment by laanwj** [comment]
You don't have to explicitly know the ports, you can use `bitcoin-cli -regtest -datadir=/tmp/<tempdir>/node0` to communicate with the instances.

File: qa/rpc-tests/test_framework.py [hunk 1]
```diff
@@ -128,10 +128,14 @@ def main(self):
             print("Unexpected exception caught during testing: "+str(e))
             traceback.print_tb(sys.exc_info()[2])
 
-        if not self.options.nocleanup:
-            print("Cleaning up")
+        try:
             stop_nodes(self.nodes)
             wait_bitcoinds()
+        except:
+            pass
```
Comment (laanwj) [inline qa/rpc-tests/test_framework.py]: Please report the exception w/ stack trace here (`traceback.print_exc()`). Ignoring exceptions, especially in tests, can cause useful diagnostic information to et lost.

Comment (dexX7) [inline qa/rpc-tests/test_framework.py]: I was thinking: if the client crashed earlier, then stopping the nodes is going to fail, too, and this was the part I wanted to hide. I'm going to take a look at it. :)

Comment (dexX7) [inline qa/rpc-tests/test_framework.py]: I might be wrong, but it looks like there doesn't seem to be an exception raised to begin with. I removed the catch completely in the updated version.

**Comment by laanwj** [comment]
utACK

**Comment by gavinandresen** [comment]
ACK

**Comment by sipa** [comment]
Concept ACK. I didn't even know --nocleanup existed.

**Comment by dgenr8** [comment]
After-the-fact ACK.  --nocleanup is indispensable when something actually breaks, which is always when developing a test, but the zombie bitcoind's were a nuisance.

**Pull request merged**